crates/composition/src/subgraphs.rs

```diff
@@ -19,6 +21,13 @@ pub struct Subgraphs {
     // Invariant: `fields` is sorted by `Field::object_id`. We rely on it for binary search.
     fields: Vec<Field>,
 
+    /// All the keys (`@key(...)`) in all the subgraphs in one container.
+    keys: Keys,
+
+    /// Storage for selection sets. Technically only a restricted version of selection sets allowed
+    /// inside `@key(fields: "...")`.
+    selection_sets: SelectionSets,
```

Comment:
why not include it inside `Keys`? The selection is specific to the key in my mind with a list of subgraphs supporting it (+ resolvable).
The structure of `Selection` could be simpler if embedded inside `Key` with something like:
```
struct Selection {
  fieldId: FieldId
  children: Vec<Selection>
}
```
Is there any benefit of merging all selections inside a BTreeSet?

Reply:
> why not include it inside Keys? The selection is specific to the key in my mind with a list of subgraphs supporting it (+ resolvable).

It may be the wrong decision because I haven't implemented these other directives yet, but I think `@provides` and `@requires` also use selection sets (https://www.apollographql.com/docs/federation/federated-types/federated-directives/#provides).

> Is there any benefit of merging all selections inside a BTreeSet?

Apart from expected smaller memory footprint and less fragmentation no. Usually there's ways to query the data structures (like range queries) that open themselves up with that kind of approaches, but it won't buy us anything in this case It's probably slower than nested vecs too, since the normal case is going to be no nesting. I'll revert to the simpler approach, I'm not convinced the shared data structure for collections is a good rabbit hole to go down into anymore.

Reply:
Yes both `@provides` and `@requires` have selection sets, but the way it's constructed currently nothing will be shared unless I miss something. 🤷 Or am I missing something? In terms of memory whether having `parentId` usize vs `children` vec probably is quite close in practice for this

Reply:
I haven't thought in details about how these two would be implemented, but I imagined something like:

```rust
struct Provides {
  selection: SelectionId
}
```

sharing the same data structure for selection sets inside Subgraphs (not that it wouldn't create new entries even for the same selection sets, indeed no sharing would happen at that level, the BTreeSet would grow large). It's orthogonal to nested vecs vs. btreeset too, we could still have a selection_sets store inside Subgraphs the same way. For now let's nest.

For the memory usage, in principle (SelectionId, SelectionId, StringId) is three words, whereas a Field struct even without parent reference is (StringId, Vec), so four words. And the BTreeSet groups allocations, the hope was that you would end with fewer allocations than selection sets in practice. But I agree at the scale we're talking about it doesn't matter until we notice it being a problem, and the reasoning isn't airtight.

I've been bitten by too many small vecs nested inside struct fields in the past so I instinctively avoid them, but in this case it's the better solution for now, it's simpler, I'm changing that module to nesting vecs inside `Key` like you suggested.

Reply:
Done